libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -62,9 +64,227 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
+
+                // Manual resize to diagonal form
+                _fix_state.velocity_covariance.resize(3);
+                _fix_state.velocity_covariance[0] = vel_acc * vel_acc;
+                _fix_state.velocity_covariance[1] = _fix_state.velocity_covariance[0];
+                _fix_state.velocity_covariance[2] = _fix_state.velocity_covariance[0];
+
+                // Manual resize to diagonal form
+                _fix2_state.covariance.resize(6);
+                _fix2_state.covariance[0] = _fix_state.position_covariance[0];
+                _fix2_state.covariance[1] = _fix_state.position_covariance[1];
+                _fix2_state.covariance[2] = _fix_state.position_covariance[2];
+                _fix2_state.covariance[3] = _fix_state.velocity_covariance[0];
+                _fix2_state.covariance[4] = _fix_state.velocity_covariance[1];
+                _fix2_state.covariance[5] = _fix_state.velocity_covariance[2];
+            }
+
+            _fix_state.height_msl_mm = loc.alt * 10;
+            _fix_state.latitude_deg_1e8 = ((uint64_t) loc.lat) * 10;
+            _fix_state.longitude_deg_1e8 = ((uint64_t) loc.lng) * 10;
+
+            _fix2_state.height_msl_mm = loc.alt * 10;
+            _fix2_state.latitude_deg_1e8 = ((uint64_t) loc.lat) * 10;
+            _fix2_state.longitude_deg_1e8 = ((uint64_t) loc.lng) * 10;
+
+            // Not saved in AP
+            //_fix_state.height_ellipsoid_mm
+
+            _fix_state.ned_velocity[0] = vel_NED.x;
+            _fix_state.ned_velocity[1] = vel_NED.y;
+            _fix_state.ned_velocity[2] = vel_NED.z;
+
+            _fix2_state.ned_velocity[0] = vel_NED.x;
+            _fix2_state.ned_velocity[1] = vel_NED.y;
+            _fix2_state.ned_velocity[2] = vel_NED.z;
+
+            _fix_state.sats_used = cgps.num_sats();
+            _fix_state.pdop = cgps.get_hdop();
+
+            _fix2_state.sats_used = _fix_state.sats_used;
+            _fix2_state.pdop = _fix_state.pdop;
+
+            switch(cgps.status()) {
+                case AP_GPS::GPS_Status::NO_GPS:
+                    _fix_state.status = uavcan::equipment::gnss::Fix::STATUS_NO_FIX;
+                    break;
+                case AP_GPS::GPS_Status::NO_FIX:
+                    _fix_state.status = uavcan::equipment::gnss::Fix::STATUS_NO_FIX;
+                    break;
+                case AP_GPS::GPS_Status::GPS_OK_FIX_2D:
+                    _fix_state.status = uavcan::equipment::gnss::Fix::STATUS_2D_FIX;
+                    break;
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_DGPS:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_RTK_FLOAT:
+                case AP_GPS::GPS_Status::GPS_OK_FIX_3D_RTK_FIXED:
+                    _fix_state.status = uavcan::equipment::gnss::Fix::STATUS_3D_FIX;
+                    break;
+            }
+            _fix2_state.status = _fix_state.status;
```

Comment:
You need to fill in the _fix2 sub_mode field for RTK float/DGPS types. (I really hate that this was taken as the solution for enumerating RTK float/etc because it makes it far more complicated to encode/decode, but there you go....

Reply:
Done